cirq-core/cirq/transformers/randomized_measurements.py

```diff
@@ -0,0 +1,109 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import Sequence
+from typing import Any, Literal
+
+import cirq
+import numpy as np
+import numpy.typing as npt
+from cirq.transformers import transformer_api
+
+
+@transformer_api.transformer
+class RandomizedMeasurements:
+    def __init__(self, *, subsystem: Sequence[int] | None = None):
+        """Class structure for performing and analyzing a general randomized measurement protocol.
+        For more details on the randomized measurement toolbox see https://arxiv.org/abs/2203.11374
+
+        Args:
+            subsystem: The specific subsystem (e.g qubit index) to measure in random basis
+        """
+        self.subsystem = subsystem
+
+    def __call__(
+        self,
+        circuit: 'cirq.AbstractCircuit',
+        rng: np.random.Generator | None = None,
+        *,
+        context: transformer_api.TransformerContext | None = None,
+    ):
+        """Apply the transformer to the given circuit. Given an input circuit returns
+        a list of circuits with the pre-measurement unitaries. If no arguments are specified,
+        it will default to computing the entropy of the entire circuit.
+
+        Args:
+            circuit: The circuit to add randomized measurements to.
+            rng: Random number generator.
+            context: Not used; to satisfy transformer API.
+
+        Returns:
+            List of circuits with pre-measurement unitaries and measurements added
+        """
+        if rng is None:
+            rng = np.random.default_rng()
+
+        qubits = sorted(circuit.all_qubits())
+        num_qubits = len(qubits)
+
+        pre_measurement_unitaries_list = self._generate_unitaries_list(rng, num_qubits)
+        pre_measurement_moment = self.unitaries_to_moment(pre_measurement_unitaries_list, qubits)
+
+        return cirq.Circuit.from_moments(
+            *circuit.moments, pre_measurement_moment, cirq.measure_each(*qubits)
```

Comment:
Maybe instead of `cirq.measure_each(*qubits)` we should use `cirq.M(*qubits, key='m')` to make it easier to extract the bitstrings to feed into `process_renyi_entropy_from_bitstrings`. (As it is, each qubit has a separate measurement key; this would give one key for the multi-qubit measurement.)